druid-shell/src/mouse.rs

```diff
@@ -18,50 +18,240 @@ use crate::kurbo::Point;
 
 use crate::keyboard::KeyModifiers;
 
-/// The state of the mouse for a click, mouse-up, or move event.
+/// Information about the mouse move event.
 #[derive(Debug, Clone, PartialEq)]
-pub struct MouseEvent {
+pub struct MoveEvent {
     /// The location of the mouse in the current window.
     ///
-    /// This is in px units, that is, adjusted for hi-dpi.
+    /// This is in px units not device pixels, that is, adjusted for hi-dpi.
     pub pos: Point,
-    /// Keyboard modifiers at the time of the mouse event.
+    /// Mouse buttons being held down at the time of the event.
+    pub buttons: MouseButtons,
+    /// Keyboard modifiers at the time of the event.
+    pub mods: KeyModifiers,
+}
+
+/// Information about the mouse click event.
+#[derive(Debug, Clone, PartialEq)]
+pub struct ClickEvent {
+    /// The location of the mouse in the current window.
+    ///
+    /// This is in px units not device pixels, that is, adjusted for hi-dpi.
+    pub pos: Point,
+    /// Mouse buttons being held down after the event.
+    /// Thus it will contain the `button` that triggered a mouse-down event,
+    /// and it will not contain the `button` that triggered a mouse-up event.
+    pub buttons: MouseButtons,
+    /// Keyboard modifiers at the time of the event.
     pub mods: KeyModifiers,
     /// The number of mouse clicks associated with this event. This will always
     /// be `0` for a mouse-up event.
-    pub count: u32,
-    /// The currently pressed button in the case of a move or click event,
-    /// or the released button in the case of a mouse-up event.
+    pub count: u8,
+    /// The button that was pressed down in the case of mouse-down,
+    /// or the button that was released in the case of mouse-up.
     pub button: MouseButton,
 }
 
 /// An indicator of which mouse button was pressed.
 #[derive(PartialEq, Eq, Clone, Copy, Debug)]
+#[repr(u8)]
 pub enum MouseButton {
     /// Left mouse button.
     Left,
-    /// Middle mouse button.
-    Middle,
     /// Right mouse button.
     Right,
+    /// Middle mouse button.
+    Middle,
     /// First X button.
     X1,
     /// Second X button.
     X2,
+    /// Some other unknown button.
+    Other,
```

Comment:
I think if we're reworking this API we should include precise arbitrary buttons, which means `Other` should be `Other(u8)`. We will end up needing to do this eventually for something, so we might as well do it now.

Reply:
Having it be `Other(u8)` would change the `MouseButtons` implementation because it couldn't just be a bitmask and would need to store arbitrary `u8`s in the set as well.

Also it doesn't really look like any of the platforms (Windows, macOS, GTK, web) support more specific buttons via the APIs that we're currently using. GTK is struggling with X1/X2 already. It's of course possible to tap into the mouse driver and get more of them, but I feel like that's a rather large project and probably not even that useful for general GUI apps.

So given the large scope of cross-platfrom support for more buttons, I think we can't reasonably prepare our API for them. When someone wants to take that task on and knows the demands of a practical implementation, then it would be much more reasonable to change the API.

Reply:
on macOS the mouse buttons are presented as a u32 bitmask. Windows seems to support up to five buttons. I haven't checked elsewhere.

Okay, so my last question would be, is `Other` even a useful distinction? If the user can't get any more information I think it could be problematic, since two separate buttons could send `Other`, and you can't tell them apart for the purposes of up & down.

In the very long term, I've been imagining we want to have some API to get the underlying event from the platform in these cases, but that's also another can of worms.